desc/compute/_neoclassical.py

```diff
@@ -0,0 +1,261 @@
+    dim=1,
+    params=[],
+    transforms={"grid": []},
+    profiles=[],
+    coordinates="r",
+    data=["B^zeta"],
+    source_grid_requirement={"coordinates": "raz", "is_meshgrid": True},
+)
+def _L_ra_fsa(data, transforms, profiles, **kwargs):
+    g = transforms["grid"].source_grid
+    shape = (g.num_rho, g.num_alpha, g.num_zeta)
+    L_ra = simpson(
+        jnp.reciprocal(data["B^zeta"]).reshape(shape),
+        jnp.reshape(g.nodes[:, 2], shape),
+        axis=-1,
+    )
+    data["<L|r,a>"] = g.expand(jnp.abs(_poloidal_mean(g, L_ra)))
+    return data
+
+
+@register_compute_fun(
+    name="<G|r,a>",
+    label="\\int_{\\zeta_{\\mathrm{min}}}^{\\zeta_{\\mathrm{max}}"
+    " \\frac{d\\zeta}{|B^{\\zeta} \\sqrt g|}",
+    units="1 / Wb",
+    units_long="Inverse webers",
+    description="(Mean) proper length over volume of field line(s)",
+    dim=1,
+    params=[],
+    transforms={"grid": []},
+    profiles=[],
+    coordinates="r",
+    data=["B^zeta", "sqrt(g)"],
+    source_grid_requirement={"coordinates": "raz", "is_meshgrid": True},
+)
+def _G_ra_fsa(data, transforms, profiles, **kwargs):
+    g = transforms["grid"].source_grid
+    shape = (g.num_rho, g.num_alpha, g.num_zeta)
+    G_ra = simpson(
+        jnp.reciprocal(data["B^zeta"] * data["sqrt(g)"]).reshape(shape),
+        jnp.reshape(g.nodes[:, 2], shape),
+        axis=-1,
+    )
+    data["<G|r,a>"] = g.expand(jnp.abs(_poloidal_mean(g, G_ra)))
+    return data
+
+
+@register_compute_fun(
+    name="effective ripple",  # this is ε¹ᐧ⁵
+    label="ε¹ᐧ⁵ = π/(8√2) (R₀/〈|∇ψ|〉)² ∫dλ λ⁻²B₀⁻¹ 〈 ∑ⱼ Hⱼ²/Iⱼ 〉",
+    units="~",
+    units_long="None",
+    description="Effective ripple modulation amplitude",
+    dim=1,
+    params=[],
+    transforms={"grid": []},
+    profiles=[],
+    coordinates="r",
+    data=[
+        "min_tz |B|",
+        "max_tz |B|",
+        "B^zeta",
+        "|B|",
+        "|B|_z|r,a",
+        "|grad(rho)|",
+        "kappa_g",
+        "<L|r,a>",
+        "R0",
+        "<|grad(rho)|>",
+    ],
+    source_grid_requirement={"coordinates": "raz", "is_meshgrid": True},
+    num_quad=(
+        "int : Resolution for quadrature of bounce integrals. Default is 31, "
+        "which gets sufficient convergence, so higher values are likely unnecessary."
+    ),
+    num_pitch=(
+        "int : Resolution for quadrature over velocity coordinate, preferably odd. "
+        "Default is 125. Effective ripple will look smoother at high values. "
+        "(If computed on many flux surfaces and micro oscillation is seen "
+        "between neighboring surfaces, increasing num_pitch will smooth the profile)."
+    ),
+    batch="bool : Whether to vectorize part of the computation. Default is true.",
+    # Some notes on choosing the resolution hyperparameters:
+    # The default settings above were chosen such that the effective ripple profile on
+    # the W7-X stellarator looks similar to the profile computed at higher resolution,
```

Comment:
would be good to check this on some other configurations (both more and less optimized) to make sure these defaults are good.

Reply:
These defaults work fine with our examples. They're chosen so that optimization works, not for high fidelity. 

The parameter that has the most significant impact is `knots_per_transit`, which is 100 by default and the user sets in the objective. If the max frequency of |B| is high than many knots are needed for local spline reconstruction of |B|. For example, Heliotron requires 200; for Heliotron, 100 significantly overestimates the effective ripple. Memory limitations prevent me from changing the default to 200. I think #1119 and #1052 will help with this.

Reply:
https://github.com/PlasmaControl/DESC/pull/1119/files#r1770596795